lib/uplink/bucket.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/storage/streams"
+	"storj.io/storj/pkg/storj"
+)
+
+// Bucket represents operations you can perform on a bucket
+type Bucket struct {
+	storj.Bucket
+
+	metainfo   *kvmetainfo.DB
+	streams    streams.Store
+	pathCipher storj.Cipher
+}
+
+// OpenObject returns an Object handle, if authorized.
+func (b *Bucket) OpenObject(ctx context.Context, path storj.Path) (o *Object, err error) {
+	defer mon.Task()(&ctx)(&err)
+	panic("TODO")
+}
+
+// UploadOptions controls options about uploading a new Object, if authorized.
+type UploadOptions struct {
+	Metadata map[string]string
```

Comment:
We should have `ContentType string` field here too.

Reply:
any reason not to just use the Metadata map for Content-Type? other systems do that

Reply:
My understanding of the `Metadata` map is that it is for used-defined or app/client-defined metadata. ContentType should be first-class metadata.

Reply:
even in a scenario where Content Type is treated specially, it's still okay to have it in the same map. many s3 libraries treat object metadata this way, even if it's user defined vs system defined. this is mainly because user defined or system defined, it's one namespace. what would it mean to have a user-defined ContentType that's different from the system defined one?

https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html

Reply:
This struct is defined by us, so the `Metadata` map can have whatever scope or meaning we want to give it. This doesn't have to map exactly to the GRPC metadata map.

That said, it doesn't seem like it matters a whole lot one way or the other.

Reply:
pff whatever this is the most important decisiokay you're right i will stop caring

Reply:
My main concern is that it is error-prone to leave this to clients. What should be the exact key for Content Type in the metadata map? `content-type`, `contenttype`, `Content-Type` or `ContentType`? These are all different keys. If two different apps choose two different keys for the Content Type, this breaks the interoperability on the same data.

Reply:
Ok, I think everybody is fine with that. I'll add a ContentType field here and in ObjectMeta.